pkg/graveler/staging/manager.go

```diff
@@ -4,112 +4,106 @@ import (
-			if err != nil {
-				return nil, err
-			}
-			if res.RowsAffected() == 0 {
-				return nil, graveler.ErrPreconditionFailed
-			}
-			return res, err
+
+	var err error
+	pb := protoFromValue(key, value)
+	if overwrite {
+		err = m.store.SetMsg(ctx, string(st), key, pb)
+	} else {
+		err = m.store.SetMsgIf(ctx, string(st), key, pb, nil)
+		if errors.Is(err, kv.ErrPredicateFailed) {
+			return graveler.ErrPreconditionFailed
 		}
-		return tx.Exec(`INSERT INTO graveler_staging_kv (staging_token, key, identity, data)
-								VALUES ($1, $2, $3, $4)
-								ON CONFLICT (staging_token, key) DO UPDATE
-									SET (staging_token, key, identity, data) =
-											(excluded.staging_token, excluded.key, excluded.identity, excluded.data)`,
-			st, key, value.Identity, value.Data)
-	}, p.txOpts()...)
+	}
 	return err
 }
 
-func (p *Manager) DropKey(ctx context.Context, st graveler.StagingToken, key graveler.Key) error {
-	_, err := p.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return tx.Exec("DELETE FROM graveler_staging_kv WHERE staging_token=$1 AND key=$2", st, key)
-	}, p.txOpts()...)
+func (m *Manager) DropKey(ctx context.Context, st graveler.StagingToken, key graveler.Key) error {
+	// Changed behavior from DB - we will not get an error on delete non-existent
+	// Correctness should maintain since if the key doesn't exist - it is still considered deleted
+	err := m.store.DeleteMsg(ctx, string(st), key)
 	return err
 }
 
+// List TODO niro: Remove batchSize parameter post KV
 // List returns an iterator of staged values on the staging token st
-func (p *Manager) List(ctx context.Context, st graveler.StagingToken, batchSize int) (graveler.ValueIterator, error) {
-	return NewStagingIterator(ctx, p.db, p.log, st, batchSize), nil
+func (m *Manager) List(ctx context.Context, st graveler.StagingToken, _ int) (graveler.ValueIterator, error) {
+	return NewStagingIterator(ctx, m.store, st)
 }
 
-func (p *Manager) Drop(ctx context.Context, st graveler.StagingToken) error {
-	_, err := p.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		return tx.Exec("DELETE FROM graveler_staging_kv WHERE staging_token=$1", st)
-	}, p.txOpts()...)
-	return err
+func (m *Manager) Drop(ctx context.Context, st graveler.StagingToken) error {
+	return m.DropByPrefix(ctx, st, []byte(""))
 }
 
-func (p *Manager) DropByPrefix(ctx context.Context, st graveler.StagingToken, prefix graveler.Key) error {
-	upperBound := graveler.UpperBoundForPrefix(prefix)
-	builder := sq.Delete("graveler_staging_kv").Where(sq.Eq{"staging_token": st}).Where("key >= ?::bytea", prefix)
-	_, err := p.db.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		if upperBound != nil {
-			builder = builder.Where("key < ?::bytea", upperBound)
-		}
-		query, args, err := builder.PlaceholderFormat(sq.Dollar).ToSql()
+func (m *Manager) DropByPrefix(ctx context.Context, st graveler.StagingToken, prefix graveler.Key) error {
+	// Wish we had 'drop partition'...
+	// Simple implementation
```

Comment:
Link to the issue where we promise to solve this better. Also - I'm not sure we want to block on it. Another option would be to flag this staging token and delete it in the background.

Reply:
This is DropByPrefix - not Drop so we can't flag the whole staging token. For Drop we might be able to do something like that, but I think we need the graveler scope for that and not staging manager. Opened https://github.com/treeverse/lakeFS/issues/3628